File: check-turn3.js
#!/usr/bin/env node

const {Dex} = require('./../pokemon-showdown-ts/dist/sim');
const {Battle} = require('./../pokemon-showdown-ts/dist/sim/battle');
const fs = require('fs');

// Get seed from command line, default to 2
const seedValue = parseInt(process.argv[2] || '2', 10);

// Load the teams we generated
const teamsFile = seedValue === 1 ? 'teams-js.json' : `teams-js-seed${seedValue}.json`;
const teams = JSON.parse(fs.readFileSync(teamsFile, 'utf8'));

const dex = Dex.mod('gen9');

// Create battle with seed
const battle = new Battle({
    formatid: 'gen9randombattle',
    seed: [0, 0, 0, seedValue],
});

// Add players with teams
battle.setPlayer('p1', {
    name: 'Player 1',
    team: teams.p1.map(p => ({
        name: p.name,
        species: p.species,
        level: p.level,
        ability: p.ability,
        item: p.item,
        nature: p.nature,
        gender: p.gender,
        moves: p.moves,
        evs: p.evs,
        ivs: p.ivs,
    })),
});

battle.setPlayer('p2', {
    name: 'Player 2',
    team: teams.p2.map(p => ({
        name: p.name,
        species: p.species,
        level: p.level,
        ability: p.ability,
        item: p.item,
        nature: p.nature,
        gender: p.gender,
        moves: p.moves,
        evs: p.evs,
        ivs: p.ivs,
    })),
});

// Run turns 1 and 2
for (let turnNum = 1; turnNum <= 2; turnNum++) {
    const logBefore = battle.log.length;
    battle.makeChoices('default', 'default');
    const turnLog = battle.log.slice(logBefore).join('\n');
    console.log(`\n=== Turn ${turnNum} Log ===`);
    console.log(turnLog);
}

// Now show detailed log for turn 3
console.log('\n=== Before Turn 3 ===');
console.log('Battle turn:', battle.turn);

const logBefore = battle.log.length;

// Make turn 3
console.log('\n=== Making Turn 3 choices ===');
battle.makeChoices('default', 'default');

console.log('\n=== After Turn 3 ===');
console.log('Battle turn:', battle.turn);

// Get only the new log entries for turn 3
const turn3Log = battle.log.slice(logBefore).join('\n');

console.log('\n=== Turn 3 Log ===');
console.log(turn3Log);

File: examples/check_turn3.rs
/// Debug turn 3 for seed 2

use pokemon_showdown::{Battle, BattleOptions, PlayerOptions, PokemonSet, PRNGSeed, ID};
use pokemon_showdown::dex_data::{StatsTable, Gender};
use std::fs;

fn main() {
    // Get seed from command line, default to 2
    let seed_value: u16 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(2);

    // Load the teams we generated
    let teams_file = if seed_value == 1 {
        "teams-rust.json".to_string()
    } else {
        format!("teams-rust-seed{}.json", seed_value)
    };
    let teams_json = fs::read_to_string(&teams_file).unwrap();

    #[derive(serde::Deserialize)]
    struct Teams {
        p1: Vec<TestPokemon>,
        p2: Vec<TestPokemon>,
    }

    #[derive(serde::Deserialize)]
    struct TestPokemon {
        name: String,
        species: String,
        level: u8,
        ability: String,
        item: String,
        nature: String,
        gender: String,
        moves: Vec<String>,
        evs: TestStats,
        ivs: TestStats,
    }

    #[derive(serde::Deserialize)]
    struct TestStats {
        hp: i32,
        atk: i32,
        def: i32,
        spa: i32,
        spd: i32,
        spe: i32,
    }

    let teams: Teams = serde_json::from_str(&teams_json).unwrap();

    let team1: Vec<_> = teams.p1.iter().map(|set| PokemonSet {
        name: set.name.clone(),
        species: set.species.clone(),
        level: set.level,
        ability: set.ability.clone(),
        item: set.item.clone(),
        nature: set.nature.clone(),
        gender: match set.gender.as_str() {
            "M" => Gender::Male,
            "F" => Gender::Female,
            _ => Gender::None,
        },
        moves: set.moves.clone(),
        evs: StatsTable::new(set.evs.hp, set.evs.atk, set.evs.def, set.evs.spa, set.evs.spd, set.evs.spe),
        ivs: StatsTable::new(set.ivs.hp, set.ivs.atk, set.ivs.def, set.ivs.spa, set.ivs.spd, set.ivs.spe),
        ..Default::default()
    }).collect();

    let team2: Vec<_> = teams.p2.iter().map(|set| PokemonSet {
        name: set.name.clone(),
        species: set.species.clone(),
        level: set.level,
        ability: set.ability.clone(),
        item: set.item.clone(),
        nature: set.nature.clone(),
        gender: match set.gender.as_str() {
            "M" => Gender::Male,
            "F" => Gender::Female,
            _ => Gender::None,
        },
        moves: set.moves.clone(),
        evs: StatsTable::new(set.evs.hp, set.evs.atk, set.evs.def, set.evs.spa, set.evs.spd, set.evs.spe),
        ivs: StatsTable::new(set.ivs.hp, set.ivs.atk, set.ivs.def, set.ivs.spa, set.ivs.spd, set.evs.spe),
        ..Default::default()
    }).collect();

    let mut battle = Battle::new(BattleOptions {
        format_id: ID::new("gen9randombattle"),
        seed: Some(PRNGSeed::Gen5([0, 0, 0, seed_value])),
        p1: Some(PlayerOptions {
            name: "Player 1".to_string(),
            team: team1,
            avatar: None,
            rating: None,
        }),
        p2: Some(PlayerOptions {
            name: "Player 2".to_string(),
            team: team2,
            avatar: None,
            rating: None,
        }),
        ..Default::default()
    });

    // Run turns 1 and 2
    for turn_num in 1..=2 {
        battle.make_choices(&["default", "default"]);
    }

    // Now show detailed log for turn 3
    println!("\n=== Before Turn 3 ===");
    println!("Battle turn: {}", battle.turn);
    println!("PRNG calls so far: {}", battle.prng.call_count);

    let log_before = battle.get_log();

    // Make turn 3
    println!("\n=== Making Turn 3 choices ===");
    battle.make_choices(&["default", "default"]);

    println!("\n=== After Turn 3 ===");
    println!("Battle turn: {}", battle.turn);
    println!("PRNG calls total: {}", battle.prng.call_count);

    // Get only the new log entries for turn 3
    let log_after = battle.get_log();
    let turn3_log = log_after[log_before.len()..].to_string();

    println!("\n=== Turn 3 Log ===");
    println!("{}", turn3_log);
}
